### PR TITLE
many: vendor apparmor into the snapd snap (take 3)

### PR DESCRIPTION
**Pull request by alexmurray**
This PR is a rework of #11096 - it updates the vendored apparmor to the 3.0.7 upstream release, plus it changes the behaviour of the apparmor sandbox in snapd to only use the vendored apparmor when the hosts snapd has a snapd-apparmor that supports reexec. Finally it extends the snapd-snap spread test to test the case where the host snapd-apparmor does not support re-exec by downgrading the snapd deb to the original version from the Ubuntu archive during the test (as this does not currently support reexec for snapd-apparmor).

### CODE REVIEW

**Comment by alexmurray** [comment]
Closing and reopening to retrigger actions checks since `unit-tests (latest/stable, normal)` seems to have unexpectedly failed.

File: sandbox/apparmor/apparmor.go
Comment (mardy) [inline sandbox/apparmor/apparmor.go:508]: you can use `testutils.Backup()`

File: sandbox/apparmor/apparmor_test.go
Comment (mardy) [inline sandbox/apparmor/apparmor_test.go:56]: I think this should work as well:
```suggestion
	c.Check(args, DeepEquals, []string{})
```

File: spread.yaml
Comment (mardy) [inline spread.yaml:234]: I guess that these changes were committed here by mistake?

File: tests/main/snapd-snap/task.yaml
Comment (mardy) [inline tests/main/snapd-snap/task.yaml:256]: please split this line

**Review by mardy — APPROVED** [review]
Thanks, mostly LGTM, but I have a few questions and nitpicks.

I'm unsure about the spread tests, because it seems we are creating a mega-test (`tests/main/snapd-snap`) to basically test that the snap package of snapd is sane, but do we really need to do that, since we already have the ubuntu-core tests which are testing the snapd snap?

File: sandbox/apparmor/apparmor.go
Comment (mardy) [reply sandbox/apparmor/apparmor.go]: Sorry, I'm afraid I gave you a wrong suggestion: we cannot use `testutils` in snapd, only in the tests; can this function be moved to `export_test.go`?

If not, then I'm afraid you should revert it to your original implementation.

File: build-aux/snap/snapcraft.yaml
Comment (pedronis) [inline build-aux/snap/snapcraft.yaml:190]: should the build spread test explicitly or does it already implicitly test this various copying?

File: spread.yaml
Comment (pedronis) [inline spread.yaml:112]: master has this already afaict I can tell, if you move the line before workers it should merge cleanly without having this twice

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:37]: we can't use testutil in the prod code itself, it brings in go-check

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:378]: this bit of code is not exercised by the unit tests

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:388]: this needs a standalone test using dirs.SetRootDir and possibly export_test.go

File: sandbox/apparmor/apparmor.go
Comment (pedronis) [inline sandbox/apparmor/apparmor.go:507]: this doesn't seem used outside of this package so it can live in export_test.go where testutil is fine

**Review by pedronis — COMMENTED** [review]
thank you, looking good. some small remarks

File: spread.yaml
Comment (alexmurray) [reply spread.yaml:234]: No this was intentional - with the changes to the snapd-snap spread test, a local qemu VM needs more memory to be able to execute it successfully.

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go]: No worries - done in  76a8d13428

File: tests/main/snapd-snap/task.yaml
Comment (alexmurray) [reply tests/main/snapd-snap/task.yaml]: Done in 8bcd2ba0ef

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go]: Yep mardy mentioned this too - I moved this to export_test in 76a8d13428

File: spread.yaml
Comment (alexmurray) [reply spread.yaml:111]: Moved in d10853b04e

File: build-aux/snap/snapcraft.yaml
Comment (alexmurray) [reply build-aux/snap/snapcraft.yaml:190]: This is implicitly tested in the snapd-snap spread test via the line: https://github.com/snapcore/snapd/pull/12087/files#diff-91d3eb78b80e15bf4e9a9f5dfe17fdda2a1e07cbb574e0113d6bea9c6bb8fb85R256

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go:387]: Added in f4c6caae95

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go]: Yep moved in 76a8d13428

**Comment by alexmurray** [comment]
> I'm unsure about the spread tests, because it seems we are creating a mega-test (`tests/main/snapd-snap`) to basically test that the snap package of snapd is sane, but do we really need to do that, since we already have the ubuntu-core tests which are testing the snapd snap?

So the reason the ubuntu-core tests are not sufficient is that they use the snapd snap from the edge channel - whereas in this case *until this PR is merged*, the snapd snap on edge does not have a vendored apparmor - and so the only way to test the snapd snap with a vendored apparmor in spread is by doing it in the snapd-snap spread test as that builds the snapd-snap directly from this branch.

**Comment by alexmurray** [comment]
@mardy @pedronis thanks for the review comments - I believe I have now addressed all of them - can you please take another look? Thanks.

File: sandbox/apparmor/apparmor_test.go
Comment (pedronis) [inline sandbox/apparmor/apparmor_test.go:314]: for cleanliness this needs to be undone with:

defer dirs.SetRootDir("")

seems there are other and older calls that needs this treatment

**Review by pedronis — APPROVED** [review]
thank you, one detail about unding dirs.SetRootDir

File: sandbox/apparmor/apparmor_test.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor_test.go:317]: Thanks - added in https://github.com/snapcore/snapd/pull/12087/commits/cef4cfecbd0d2f673151ba9cb87a258d85e07d83

**Review by mardy — APPROVED** [review]
Thanks for the explanations, LGTM!

**Comment by alexmurray** [comment]
@mvo5 this has approvals from mardy and pedronis - can it be merged? Thanks.

File: tests/nightly/security-internal-apparmor/task.yaml
Comment (mvo5) [inline tests/nightly/security-internal-apparmor/task.yaml:1]: Silly question - what is the benefit of this test? Could we test the same in the `snapd-snap` test?

**Review by mvo5 — COMMENTED** [review]
Thanks, some questions inline but this looks super close!

File: sandbox/apparmor/profile.go
Comment (alexmurray) [reply sandbox/apparmor/profile.go:103]: Since this fallback is only required when doing unit testing to mock the parser command, I think perhaps I'll remove it entirely and update the unit tests to make sure they all MockParserSearchPath() as expected and then hopefully it "just works"

File: tests/nightly/security-internal-apparmor/task.yaml
Comment (alexmurray) [reply tests/nightly/security-internal-apparmor/task.yaml:1]: This predates the `snapd-snap` test - and infact we already do this and more in the `snapd-snap` test - I will remove it.

File: build-aux/snap/apparmor/af_names.h
Comment (alexmurray) [reply build-aux/snap/apparmor/af_names.h:2]: Thanks - I will move it to build-aux/snap/local/apparmor/af_names.h then

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go:395]: Sounds good - I'll change it to return a `exec.Cmd`

File: interfaces/apparmor/backend.go
Comment (alexmurray) [reply interfaces/apparmor/backend.go:872]: Sounds good.

Comment (alexmurray) [reply interfaces/apparmor/backend.go]: See bb7af271bc

File: sandbox/apparmor/apparmor.go
Comment (alexmurray) [reply sandbox/apparmor/apparmor.go]: See fd767ee5af

File: sandbox/apparmor/profile.go
Comment (alexmurray) [reply sandbox/apparmor/profile.go]: See fd767ee5af

File: build-aux/snap/apparmor/af_names.h
Comment (alexmurray) [reply build-aux/snap/apparmor/af_names.h]: See e832daf89e

File: tests/nightly/security-internal-apparmor/task.yaml
Comment (alexmurray) [reply tests/nightly/security-internal-apparmor/task.yaml]: See dce92ac73e

**Comment by alexmurray** [comment]
Closing and reopening to retrigger github actions since failed previously on a proxy error trying to contact launchpad.net when building the snapd snap.

**Comment by alexmurray** [comment]
Ping @mvo5 - I see you already self-assigned to review this but just want to make sure it doesn't get forgotten :) thanks

**Review by mvo5 — APPROVED** [review]
Thanks, this looks good!

**Comment by alexmurray** [comment]
@mvo5 I just pushed a fix for the failing snapd-snap spread test on arm64

**Review by mardy — APPROVED** [review]
Re-approving, the last commits are fine too.

**Comment by alexmurray** [comment]
@mvo5 is there anything you need from me to get this merged?

**Comment by valentindavid** [comment]
@alexmurray @mvo5 I cannot build the snap of snapd locally anymore since this PR.

Was something changed in the way to build the snap? I have tried using snapcraft 4.x and do a clean build. There are missing build dependencies for apparmor.

**Comment by valentindavid** [comment]
Opened #12348 to fix it.

**Pull request merged**
